Sources/Fuzzilli/Lifting/TypeCollectionAnalyzer.swift

```diff
@@ -0,0 +1,36 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Determine what variable types we should collect given instruction
+struct TypeCollectionAnalyzer {
+    // Special properties which significantly change variable type
+    // We should recollect type if we change them
+    let specialProperties = ["__proto__"]
+
+    func analyze(_ instr: Instruction) -> [Variable] {
+        switch instr.operation {
```

Comment:
Ah ok I see. I was initially suggesting to only let this Analyzer handle **re**collection cases and still have the JSLifter collect types for all outputs. But I think this is actually cooler! Because now all the logic about when to collect type information is in a single place :)
I'm wondering though, can you do something like this instead:
```
swift instr.operation {
case is LoadInt, is LoadBigInt, is LoadString, ...:
    // No need to collect types for literals. The interpreter can handle them
    return []
case is BeginAnyFunctionDefinition:
    // We don't collect parameter types ... (actually, I think maybe we should, but let's discuss some other time :D)
    return ...
case let op as StoreProperty where ...:
    variablesForCollection.append(instr.input(0))
default:
    // By default, collect types for all outputs of an instruction
    return Array(instruction.allOutputs)
}
return variablesForCollection
}
```
Then it's obvious what *isn't* collected, and it's also fail safe for the future, as by default, everything will be collected. WDYT?

Reply:
The question is what should be the default behavior.
1. optimistic approach (you suggest) - type collection of newly implemented instruction nearly always result in gathering all output types and it is correctly lifted in usual way (after lifting instruction, lift `updateType` with all output variables one by one). We already catched special cases and if new instruction is somehow special programmer should know about this and fix it himself (otherwise unexpected behavior)

2. conservative approach (originally implemented) - When new instruction is implemented, programmer has to update this function to say how we should collect type information for this new instruction explicitely. If programmer forgets, there is no unexpected behavior, just no benefits from type collecting new instruction

I don't have strong opinion here, so can go with your one